index.js

```diff
@@ -1,102 +1,73 @@
 'use strict';
-const fs = require('fs');
-const path = require('path');
 const {
   getBindOptions,
   setKey,
-  getKeyMapping,
-  getValueMapping
+  mapKey,
+  getBindingData,
+  mapValue,
+  isKnownServiceType,
+  getClientInfo,
+  isDefined,
+  getBindingDataPath,
+  getRawBindingData
 } = require('./utils/index.js');
 
-const typeMapping = {
-  KAFKA: 'kafka',
-  POSTGRESQL: 'postgresql',
-  REDIS: 'redis',
-  MONGODB: 'mongodb',
-  AMQP: 'amqp',
-  MYSQL: 'mysql'
-};
-
-const aliases = {
-  amqp: ['rabbitmq']
-};
+const {
+  errors: {
+    NO_SERVICE_BINDING_ROOT,
+    UNKNOWN_SERVICE_TYPE,
+    UNKNOWN_CLIENT,
+    NO_BINDING_FOUND
+  }
+} = require('./utils/messages/index.js');
 
-// return the bindings requested
 function getBinding(type, client, bindingOptions) {
-  const bindOptions = getBindOptions(bindingOptions);
-
-  const id = bindOptions.id;
+  const bindOptions = getBindOptions(type, client, bindingOptions);
 
-  // validate we know about the type
-  if (!fs.existsSync(path.join(__dirname, 'clients', type))) {
-    throw new Error('Unknown service type');
-  }
+  const root = process.env.SERVICE_BINDING_ROOT;
 
-  // validate that we know about the requested client
-  let clientInfo;
-  if (client) {
-    try {
-      clientInfo = require(path.join(__dirname, 'clients', type, client));
-    } catch (e) {
-      throw new Error('Unknown client');
+  if (!isDefined(type)) {
+    if (!root) {
+      throw new Error(NO_SERVICE_BINDING_ROOT);
     }
+    return getRawBindingData(root);
   }
 
-  // find the matching binding
-  let bindingsRoot = null;
-  const root = process.env.SERVICE_BINDING_ROOT;
-
-  if (!root) {
-    throw new Error('No SERVICE_BINDING_ROOT env variable Found');
+  // validate we know about the type
+  if (isDefined(type) && !isKnownServiceType(type)) {
+    throw new Error(UNKNOWN_SERVICE_TYPE);
   }
 
-  const candidates = fs.readdirSync(root);
-  for (const file of candidates) {
-    try {
-      const bindingType = fs
-        .readFileSync(path.join(root, file, 'type'))
-        .toString()
-        .trim();
-      if (
-        bindingType === typeMapping[type] ||
-        aliases[typeMapping[type]].includes(bindingType)
-      ) {
-        if (id === undefined || file.includes(id)) {
-          bindingsRoot = path.join(root, file);
-          break;
-        }
-      }
-    } catch (err) {}
+  // validate we know about the client
+  const clientInfo = getClientInfo(type, client);
+  if (isDefined(client) && !clientInfo) {
+    throw new Error(UNKNOWN_CLIENT);
   }
 
-  // bail if we did not find a binding
-  if (bindingsRoot == null) {
-    throw new Error('No Binding Found');
+  let bindingData;
+  if (isDefined(bindOptions.bindingData)) {
+    bindingData = Object.entries(bindOptions.bindingData);
+  } else {
+    if (!root) {
+      throw new Error(NO_SERVICE_BINDING_ROOT);
+    }
```

Comment:
Can we do this check once right after the line with `const root = process.env.SERVICE_BINDING_ROOT;` and throw the error there if there is no service binding right at the start versus having to check in a few places?

Reply:
yeap!

Reply:
90f777e95ae4d6f6fc21e6fdb315969d4990b46d